core/blockchain.go

```diff
@@ -1240,13 +1241,18 @@ func (bc *BlockChain) WriteBlockWithState(
 			}
 
 			members := []*bls2.PublicKey{}
+			addrs, isNewEpoch := []common.Address{}, len(block.Header().ShardState()) > 0
+
 			for _, slot := range committee.Slots {
 				pubKey := &bls2.PublicKey{}
 				err := pubKey.Deserialize(slot.BlsPublicKey[:])
 				if err != nil {
 					utils.Logger().Err(err).Msg("[Uptime] Failed to deserialize bls public key")
 				}
 				members = append(members, pubKey)
+				if isNewEpoch {
+					addrs = append(addrs, slot.EcdsaAddress)
```

Comment:
You can directly call ReadActiveValidatorList to get all the validators of all shards at this epoch. And I think you can do the SetInactiveUnavailableValidators once for all validators right? Also don't need to embed this logic in crosslinks, just do it separately.